catlearn/relation_cache.py

```diff
@@ -390,11 +390,11 @@ def build_composites(
 
             added_arrows = set()
             # loop through length n arrows and try to extend them by 1 node
-            for arr in self.arrows(
+            for arr in set(self.arrows(
```

Comment:
Cache starts build after the fix with `max_arrow_number` provided. However it times out with even max_arrow_number = 1 when build against graph_train.
Basically 16Gb of free RAM are taken after some time and the machine is 'frozen'.

Reply:
@Kuraao That was the bug you told me about? Looks like it from the looks (adding edges while looping). As you point out @igorgarbuz the fix seems to blow up the memory.

Since you only add extended arrow, it semantically does not change the set of arrows we iterate on. So you could keep the generators as before, only adding to `added_arrows` and then after the nested for loops are done, you can add them all - right before `prune_relations()`

This should reduce memory usage (not having to store the sets of arrows) and correct the bug.